src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts

```diff
@@ -411,20 +411,22 @@ class RemoteMachineTrainingService implements TrainingService {
         const rmMetaList: RemoteMachineMeta[] = <RemoteMachineMeta[]>JSON.parse(machineList);
         let connectedRMNum: number = 0;
 
-        rmMetaList.forEach(async (rmMeta: RemoteMachineMeta) => {
+        const connectionPromises = [];
+        for (const rmMeta of rmMetaList) {
             rmMeta.occupiedGpuIndexMap = new Map<number, number>();
             const executorManager: ExecutorManager = new ExecutorManager(rmMeta);
             this.log.info(`connecting to ${rmMeta.username}@${rmMeta.ip}:${rmMeta.port}`);
             const executor: ShellExecutor = await executorManager.getExecutor(this.initExecutorId);
             this.log.debug(`reached ${executor.name}`);
             this.machineExecutorManagerMap.set(rmMeta, executorManager);
             this.log.debug(`initializing ${executor.name}`);
-            await this.initRemoteMachineOnConnected(rmMeta, executor);
+            connectionPromises.push(this.initRemoteMachineOnConnected(rmMeta, executor));
             this.log.info(`connected to ${executor.name}`);
             if (++connectedRMNum === rmMetaList.length) {
                 deferred.resolve();
             }
-        });
+        }
+        Promise.all(connectionPromises);
```

Comment:
This new logic seems confusing when it works together with `++connectedRMNum` and `deferred.resolve()`.
Missing `await` before `Promise.all`

Reply:
fixed